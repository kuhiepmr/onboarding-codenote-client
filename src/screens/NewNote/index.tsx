import React, { Component } from "react";
import { ControlLabel, FormControl, FormGroup } from 'react-bootstrap';
import LoaderButton from '../../components/LoaderButton';
import config from '../../config';
import { s3Upload } from '../../libs/awsLib';
import { createNote } from "../../services/notes";
import { formatBytes } from '../../utils/format'
import './index.css';

class NewNote extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: null,
      content: "",
    };
  }

  isValidForm = () => (this.state.content.length)

  formatFilename(str) {
    return str.replace(/^\w+-/, "");
  }

  handleFieldChange = ({ target }) => {
    const { id, value } = target;

    this.setState({
      [id]: value
    });
  };

  handleFileChange = event => {
    const file = event.target.files[0];

    if (file && file.size > config.MAX_ATTACHMENT_SIZE) {
      alert(`Please pick a file smaller than ${formatBytes(config.MAX_ATTACHMENT_SIZE)}`);
      // remove invalid file from input
      event.target.value = null;
      return;
    }

    this.file = file;
  }

  handleSubmit = async event => {
    event.preventDefault();

    this.setState({ isLoading: true });

    const { content } = this.state;

    try {
      const payload = {
        content,
      }

      if (this.file) {
        payload.attachment = await s3Upload(this.file);
      }

      await createNote(payload)

      this.props.history.push("/");
    } catch (e) {
      alert(e);
      this.setState({ isLoading: false });
    }
  }

  render() {
    const { content, isLoading } = this.state;

    return (
      <div className="NewNote">
        <form onSubmit={this.handleSubmit}>
          <FormGroup controlId="content">
            <FormControl
              onChange={this.handleFieldChange}
              value={content}
              componentClass="textarea"
            />
          </FormGroup>
          <FormGroup controlId="file">
            <ControlLabel>Attachment</ControlLabel>
            <FormControl onChange={this.handleFileChange} type="file" />
          </FormGroup>
          <LoaderButton
            block
            bsStyle="primary"
            bsSize="large"
            disabled={!this.isValidForm()}
            type="submit"
            isLoading={isLoading}
            text="Create"
            loadingText="Creating..."
          />
        </form>
      </div>
    );
  }
}

export default NewNote
