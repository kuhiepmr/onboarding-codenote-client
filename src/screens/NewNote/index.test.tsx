import { shallow } from 'enzyme';
import React from 'react';
import configureStore from 'redux-mock-store';
import NewNote from './index';
import { History } from 'history';

describe('NewNote screen', () => {
  let wrapper, instance
  const props = {
    history: {} as History
  }

  beforeEach(() => {

    wrapper = shallow(<NewNote {...props} />);
    instance = wrapper.instance();
  });

  it('should render success', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should change field and validate', () => {
    const contentEvent = {
      target: {
        id: 'content',
        value: "test"
      }
    }

    const fileEvent = {
      target: {
        files: [
          'test'
        ]
      }
    }

    instance.handleChange(contentEvent);
    instance.handleFileChange(fileEvent);

    expect(wrapper.state().content).toEqual('test');
    expect(instance.file).toEqual(fileEvent.target.files[0]);
    expect(instance.validateForm()).toBe(true);
  });


  it('should loading when submit', () => {
    instance.handleSubmit({ preventDefault: () => { } })
    expect(wrapper.state().isLoading).toBe(true);
  });
});