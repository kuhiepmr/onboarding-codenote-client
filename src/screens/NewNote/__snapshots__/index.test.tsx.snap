// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewNote screen should render success 1`] = `
<div
  className="NewNote"
>
  <form
    onSubmit={[Function]}
  >
    <FormGroup
      bsClass="form-group"
      controlId="content"
    >
      <FormControl
        bsClass="form-control"
        componentClass="textarea"
        onChange={[Function]}
        value=""
      />
    </FormGroup>
    <FormGroup
      bsClass="form-group"
      controlId="file"
    >
      <ControlLabel
        bsClass="control-label"
        srOnly={false}
      >
        Attachment
      </ControlLabel>
      <FormControl
        bsClass="form-control"
        componentClass="input"
        onChange={[Function]}
        type="file"
      />
    </FormGroup>
    <LoaderButton
      block={true}
      bsSize="large"
      bsStyle="primary"
      disabled={true}
      isLoading={false}
      loadingText="Creatingâ€¦"
      text="Create"
      type="submit"
    />
  </form>
</div>
`;
